File: python/src/st_game.py
#
# STAR TREK Text Game
# 2017-05-01 K.OHWADA
#
	
## class
class Game():
	
	SUPPLY_ENERGY = 3000
	SUPPLY_TORPEDO = 10
	
	ENERGY_WARP = 10
	
	quadrant = None
	sector = None

	days_left = 0
	days_elapsed = 0
	
	energy = 0
	shield = 0
	torpedo = 0
	is_docked = False
	
### init	
	def __init__(self):
		from st_quadrant import Quadrant
		from st_sector import Sector
		self.quadrant = Quadrant()
		self.sector = Sector()
#		self.game_input = Input()
#		self.device = Device()
				
### def end

	def setup(self):

		self.quadrant.setup_position()
		self.quadrant.setup_q()
		self.quadrant.count_q()
		q = self.quadrant.get_current_q()
		self.sector.setup( q )

		
		self.days_left = 300 + 2 *	self.quadrant.num_klingon
		self.days_elapsed = 0				
		self.energy = self.SUPPLY_ENERGY
		self.torpedo = self.SUPPLY_TORPEDO
		self.shield = 0
						
### def end
			
	def print_long_sensor(self, is_computer_available):
		self.quadrant.print_long_sensor( is_computer_available )
## def


	def print_short_sensor(self):
		self.sector.print_short_sensor()
## def	
		
	def warp(self, course, distance ):
		if ( course <1 ) or (  course > 8 ) or ( distance < 1 ):
			return;

		self.quadrant.warp( course, distance )
		q = self.quadrant.get_current_q()
		self.sector.setup( q )
		self.decrease_energy( self.ENERGY_WARP*distance )
		self.elapse_days( distance )		
### def end	

	def elapse_days(self, d):
		self.days_left -= d
		self.days_elapsed += d
		if ( self.days_left < 0 ):
			self.days_left = 0			
### def end	

	def move(self, course, distance ):
		if ( course < 1) or ( course > 8 ) or ( distance < 1 ):
			return;
			
		self.is_docked = False
		code = self.sector.move( course, distance )
		if ( code == self.sector.C_STARBASE ):
			self.is_docked = True
		self.save_num_q()
		return code
### def end	

	def fire_phaser(self, energy ):
		self.decrease_energy( energy ) 
		self.sector.fire_phaser(energy) 
		self.save_num_q()
### def end	
			
	def fire_torpedo(self, course ):
		if ( course <1 ) or ( course > 8 ):
			return;

		self.sector.fire_torpedo(course)
		self.save_num_q()
### def end	

	def print_galaxy_record(self):
		self.quadrant.print_galaxy_record()
## def

	def print_torpedo_data(self):
		self.sector.print_torpedo_data()
## def

	def fight_back(self):
		list = self.sector.fight_back()
		if ( len(list) == 0):
			return
			
		for param in list:
			x = param[0]
			y  = param[1]
			beam = param[2]
			self.decrease_shield( beam )
			msg = " %d UNIT HIT ON ENTERPRISE AT SECTOR %d , %d ( %d LEFT)"% ( beam, x, y, self.shield )
			print msg
							
### def end	

	def decrease_shield(self, s):
		self.shield -= s
		if ( self.shield < 0 ): 
			self.shield = 0
### def end	
						
	def shield_control(self, energy ):
		self.decrease_energy( energy ) 
		self.shield = energy 
		print "SHIELD = " + str(energy)
### def end

	def decrease_energy(self, energy):
		self.energy -= energy 
		if (self.energy < 0):
			self.energy = 0
### def end

	def save_num_q(self):
		num1 = self.sector.get_num()
		self.quadrant.save_current_q(num1)
		self.quadrant.count_q()

	### def end	
	
	def dockin(self):	
		print "SHIELDS DROPPED FOR DOCKING PURPOSES"
		self.energy = self.SUPPLY_ENERGY
		self.torpedo = self.SUPPLY_TORPEDO
		self.shield = 0	
	### def end	
							
		### class end

File: javascript/src/js/st_game.js
//
// STAR TREK Text Game
// 2017-06-01 K.OHWADA
//

var Game = function () {
	
	this.quadrant = new Quadrant();
	this.sector = new Sector();

	this.days_left = 0;
	this.days_elapsed = 0;
	
	this.energy = 0;
	this.shield = 0;
	this.torpedo = 0;
	this.is_docked = false;

} // constructor


Game.prototype.setup = function () {

		this.quadrant.setup_position();
		this.quadrant.setup_q();
		this.quadrant.count_q();
		var q  = this.quadrant.get_current_q();
		this.sector.setup_s( q );

		
		this.days_left = 300 + 2 *	this.quadrant.num_klingon ;
		
		this.days_elapsed = 0 ;				
		this.energy = 3000 ; 
		this.torpedo = 10 ;
		this.shield = 0 ;
						
} // setup

Game.prototype.long_sensor = function (is_computer_available) {
				
		return this.quadrant.get_long_map( is_computer_available );

} // long_sensor


Game.prototype.short_sensor = function () { 

		return this.sector.get_short_map();
		
} // short_sensor	


Game.prototype.galaxy_record = function () {
				
		return this.quadrant.get_galaxy_map();

} // galaxy_record


Game.prototype.warp = function (course) {
//	alert("game warp");
		if (( course <1 ) || (  course > 8 ) ) {
			return;
		} // if
		
		this.quadrant.warp( course );
		var q = this.quadrant.get_current_q();
		this.sector.setup_s( q );
		this.decrease_energy( 20 );
		this.elapse_days( 1 );
		var msg = "arrive at Quadrant " +  this.quadrant.qx + " , " +this.quadrant.qy;
		alert(msg);		
} // warp


Game.prototype.elapse_days = function (d) { 

		this.days_left -= d;
		this.days_elapsed += d;
		if ( this.days_left < 0 ) {
			this.days_left = 0 ;			
		} // if
		
} // elapse_days


Game.prototype.move = function (course) { 

		if (( course < 1) || ( course > 8 ) ) {
			return;
		} // if
		
		this.is_docked = false ;
		
		var code = this.sector.move( course );  // todo
		
		if ( code == this.sector.C_STARBASE ) {
			self.is_docked = true ;
	} // if
			this.decrease_energy( 10 );	
		this.save_num_q();
		return code ;

} // move


Game.prototype.fire_phaser = function (e) { 

		this.decrease_energy(e);
		 
		var list = this.sector.fire_phaser();  // todo
		
		this.save_num_q();
		
		return list;
		
} // fire_phaser 


Game.prototype.fire_torpedo = function (course) { 			
		if (( course <1 ) && ( course > 8 )) {
			return;
		} // if
		
		var list = this.sector.fire_torpedo(course);
		this.save_num_q();
		
		return list;
		
} // fire_torpedo


Game.prototype.galaxy_record = function () { 

		return this.quadrant.get_galaxy_map();
} // galaxy_record


Game.prototype.torpedo_data = function () { 

	alert("game torpedo_data");

		return this.sector.torpedo_data();
		
} // torpedo_data		


Game.prototype.fight_back = function () { 

		return this.sector.fight_back();
		
} // fight_back		


Game.prototype.decrease_shield = function (s) { 

		this.shield -= s;
		if ( this.shield < 0 ){ 
			this.shield = 0;
	} // if
} // decrease_shield

		
Game.prototype.shield_control = function (e) { 						
		this.decrease_energy(e ); 
		this.shield = e; 
//		alert( "shield on" );

} // shield_control		


Game.prototype.decrease_energy = function (e) { 

		this.energy -= e ;
		if ( this.energy < 0) {
			this.energy = 0;
	} // if	
				
} // decrease_nergy
		 			

Game.prototype.save_num_q = function () { 

		var num = this.sector.get_num();
		this.quadrant.save_current_q(num);
		this.quadrant.count_q();

} // save_num_q
	
	
	Game.prototype.dockin = function () { 	
		this.energy = 3000;
		this.torpedo = 10;
		this.shield = 0	;

} // dockin
